# Modeling

The purpose of this notebook is to develop and evaluate machine learning models capable of predicting next year's California Dashboard color classification for Aspire Public Schools.

The preprocessing notebook prepared the dataset by:
- creating the future prediction target,
- engineering features,
- encoding categorical variables,
- standardizing numeric variables,
- and splitting the data into training and testing datasets.

This notebook will:
- train multiple classification models,
- evaluate model performance,
- compare results across models,
- and identify the final model for the project.

In [1]:
# Import libraries

import pandas as pd
import numpy as np

from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier

from sklearn.metrics import (
    accuracy_score,
    classification_report,
    confusion_matrix
)

In [2]:
# Load preprocessed datasets

X_train = pd.read_csv("X_train_scaled.csv")
X_test = pd.read_csv("X_test_scaled.csv")

y_train = pd.read_csv("y_train.csv").squeeze()
y_test = pd.read_csv("y_test.csv").squeeze()

# Check dataset shapes

print("X_train:", X_train.shape)
print("X_test:", X_test.shape)

print("y_train:", y_train.shape)
print("y_test:", y_test.shape)

X_train: (213, 7)
X_test: (71, 7)
y_train: (213,)
y_test: (71,)


## Baseline Model: Logistic Regression

The first model used in this project is Logistic Regression. Logistic Regression is commonly used as a baseline classification model because it is relatively simple, interpretable, and computationally efficient.

This model will provide a benchmark for comparing the performance of more complex ensemble models later in the project.

In [3]:
# Create Logistic Regression model

log_reg = LogisticRegression(
    max_iter=5000,
    multi_class='multinomial'
)

# Fit model on training data

log_reg.fit(X_train, y_train)

# Generate predictions

y_pred_log = log_reg.predict(X_test)

C:\Users\hillk\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:1247: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. From then on, it will always use 'multinomial'. Leave it to its default value to avoid this warning.
  warnings.warn(


In [4]:
# Evaluate Logistic Regression model

print("Logistic Regression Accuracy:")
print(accuracy_score(y_test, y_pred_log))

print("\nClassification Report:")
print(classification_report(y_test, y_pred_log))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_log))

Logistic Regression Accuracy:
0.36619718309859156

Classification Report:
              precision    recall  f1-score   support

           1       0.33      0.07      0.12        14
           2       0.23      0.17      0.19        18
           3       0.37      0.71      0.49        24
           4       0.25      0.14      0.18         7
           5       0.80      0.50      0.62         8

    accuracy                           0.37        71
   macro avg       0.40      0.32      0.32        71
weighted avg       0.36      0.37      0.32        71


Confusion Matrix:
[[ 1  6  6  1  0]
 [ 1  3 14  0  0]
 [ 1  3 17  2  1]
 [ 0  0  6  1  0]
 [ 0  1  3  0  4]]


## Logistic Regression Results

The Logistic Regression model achieved an accuracy of approximately 36.6% on the testing dataset.

The model performed best when predicting Dashboard color class 3, but struggled to accurately classify several of the smaller classes, particularly classes 1, 2, and 4.

This outcome suggests that the relationship between the predictor variables and Dashboard color classifications may be nonlinear and more complex than Logistic Regression can effectively capture.

Because Logistic Regression is a relatively simple linear model, it serves primarily as a useful baseline for comparison against more advanced ensemble methods.

## Model 2: Random Forest Classifier

The next model used is a Random Forest Classifier.

Random Forest is an ensemble learning method that combines many decision trees to improve predictive performance and reduce overfitting. Unlike Logistic Regression, Random Forest can capture nonlinear relationships and interactions between variables.

This model may perform better on the Aspire Dashboard dataset because educational performance outcomes are likely influenced by complex combinations of historical indicators and trends.

In [5]:
# Create Random Forest model

rf_model = RandomForestClassifier(
    n_estimators=200,
    max_depth=5,
    random_state=42
)

# Fit model

rf_model.fit(X_train, y_train)

# Generate predictions

y_pred_rf = rf_model.predict(X_test)

In [6]:
# Evaluate Random Forest model

print("Random Forest Accuracy:")
print(accuracy_score(y_test, y_pred_rf))

print("\nClassification Report:")
print(classification_report(y_test, y_pred_rf))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_rf))

Random Forest Accuracy:
0.3380281690140845

Classification Report:
              precision    recall  f1-score   support

           1       0.40      0.14      0.21        14
           2       0.14      0.11      0.12        18
           3       0.37      0.62      0.46        24
           4       0.33      0.14      0.20         7
           5       0.50      0.50      0.50         8

    accuracy                           0.34        71
   macro avg       0.35      0.30      0.30        71
weighted avg       0.33      0.34      0.31        71


Confusion Matrix:
[[ 2  6  6  0  0]
 [ 2  2 13  0  1]
 [ 1  4 15  2  2]
 [ 0  0  5  1  1]
 [ 0  2  2  0  4]]


## Random Forest Results

The Random Forest model achieved an accuracy of approximately 33.8%, which was slightly lower than the Logistic Regression baseline model.

Like Logistic Regression, the Random Forest model performed best when predicting Dashboard color class 3, while struggling with several of the smaller classes.

Although Random Forest is capable of modeling nonlinear relationships, the relatively small dataset size may limit the model’s ability to generalize effectively. Ensemble tree models often perform best with larger training datasets.

Despite the lower overall accuracy, Random Forest remains valuable because it can capture more complex feature interactions and may provide useful feature importance insights later in the analysis.

## Model 3: Gradient Boosting Classifier

The final model evaluated is a Gradient Boosting Classifier.

Gradient Boosting builds trees sequentially, with each new tree attempting to correct the errors made by previous trees. This often produces stronger predictive performance than individual decision trees or standard Random Forest models.

Because Gradient Boosting can model complex nonlinear relationships while reducing bias, it is frequently one of the strongest-performing algorithms for structured tabular datasets.

In [7]:
# Create Gradient Boosting model

gb_model = GradientBoostingClassifier(
    n_estimators=100,
    learning_rate=0.05,
    max_depth=3,
    random_state=42
)

# Fit model

gb_model.fit(X_train, y_train)

# Generate predictions

y_pred_gb = gb_model.predict(X_test)

In [8]:
# Evaluate Gradient Boosting model

print("Gradient Boosting Accuracy:")
print(accuracy_score(y_test, y_pred_gb))

print("\nClassification Report:")
print(classification_report(y_test, y_pred_gb))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_gb))

Gradient Boosting Accuracy:
0.36619718309859156

Classification Report:
              precision    recall  f1-score   support

           1       0.33      0.14      0.20        14
           2       0.21      0.22      0.22        18
           3       0.46      0.67      0.54        24
           4       0.29      0.29      0.29         7
           5       0.50      0.25      0.33         8

    accuracy                           0.37        71
   macro avg       0.36      0.31      0.32        71
weighted avg       0.36      0.37      0.34        71


Confusion Matrix:
[[ 2  7  5  0  0]
 [ 2  4 10  2  0]
 [ 1  5 16  1  1]
 [ 1  0  3  2  1]
 [ 0  3  1  2  2]]


## Gradient Boosting Results

The Gradient Boosting model achieved an accuracy of approximately 36.6%, matching the Logistic Regression model and outperforming the Random Forest model slightly.

Gradient Boosting showed improved balance across several Dashboard color classes compared to the previous models, particularly for classes 2 and 4. The model again performed best when predicting Dashboard color class 3, which was also the largest class in the dataset.

Overall, the results suggest that the dataset contains meaningful predictive signal, but the relatively small sample size and multiclass structure make accurate prediction challenging.

Because Gradient Boosting can capture nonlinear relationships and interactions between variables while maintaining competitive accuracy, it appears to be one of the strongest candidate models for this project.

# Compare Model Performance

The three classification models produced relatively similar performance results overall.

| Model | Accuracy |
|---|---|
| Logistic Regression | 36.6% |
| Random Forest | 33.8% |
| Gradient Boosting | 36.6% |

Logistic Regression and Gradient Boosting achieved the highest accuracy scores, while Random Forest performed slightly worse on the testing dataset.

Although Logistic Regression matched Gradient Boosting in overall accuracy, Gradient Boosting is likely the stronger final model because:
- it better captures nonlinear relationships,
- it handles feature interactions more effectively,
- and it demonstrated somewhat more balanced performance across multiple Dashboard color classes.

The relatively modest performance across all models is likely influenced by:
- the small dataset size,
- multiclass prediction difficulty,
- and overlap between Dashboard performance categories.

# Final Model Selection

The final model selected for this project is the Gradient Boosting Classifier.

Gradient Boosting was selected because it:
- achieved competitive accuracy,
- handled multiclass prediction reasonably well,
- captured nonlinear relationships between educational performance indicators,
- and demonstrated more balanced classification behavior than the other evaluated models.

While overall predictive performance remains moderate, the model still provides valuable insight into the relationship between historical Dashboard indicators and future school accountability outcomes.

Future improvements could include:
- expanding the dataset with additional years of Dashboard data,
- incorporating subgroup-level observations,
- adding additional school demographic or operational features,
- and performing more extensive hyperparameter tuning.

# Conclusion

In this notebook, I developed and evaluated multiple machine learning classification models to predict next year's California Dashboard color classifications for Aspire Public Schools.

The models included:
- Logistic Regression
- Random Forest
- Gradient Boosting

After comparing model performance using accuracy, precision, recall, F1-score, and confusion matrices, Gradient Boosting was selected as the final model for the project.

The results suggest that historical Dashboard performance metrics contain meaningful predictive information about future accountability outcomes, although prediction remains challenging due to the relatively small dataset and multiclass nature of the problem.

This modeling process demonstrates how machine learning techniques can be applied to educational accountability data to support forecasting and decision-making within school systems.